File: Secure Code Review/XSS/Java/storexss.java
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.*;
import java.io.*;
import java.util.ArrayList;
import java.util.List;

@WebServlet("/comments")
public class VulnerableStoredXSSServlet extends HttpServlet {
    private static List<String> comments = new ArrayList<>(); // Stores user-submitted comments

    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        String comment = request.getParameter("comment"); // User-supplied input
        comments.add(comment); // Store the comment without sanitization

        response.sendRedirect("/comments"); // Redirect to view comments
    }

    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
        response.setContentType("text/html");
        PrintWriter out = response.getWriter();

        out.println("<html><body>");
        out.println("<h1>Comments</h1>");
        out.println("<form method='POST' action='/comments'>");
        out.println("Comment: <input type='text' name='comment' />");
        out.println("<button type='submit'>Submit</button>");
        out.println("</form>");
        out.println("<ul>");

        // Display stored comments without escaping user input
        for (String comment : comments) {
            out.println("<li>" + comment + "</li>"); // Vulnerable to stored XSS
        }

        out.println("</ul>");
        out.println("</body></html>");
    }
}


















////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
.................................................................................................................................................................
  

How This Code is Vulnerable

    User Input is Stored and Displayed Without Validation or Escaping:
        The comment parameter is directly added to the comments list and then displayed on the page without any sanitization or escaping.
        If an attacker submits a malicious comment (e.g., <script>alert('XSS')</script>), it is stored and executed every time the comments section is loaded.

    Example of Exploitation:
        An attacker submits the following comment:

    <script>alert('XSS')</script>

    Every user who visits the comments page will see an alert pop-up triggered by the malicious script.

Impact:

    Malicious scripts could:
        Steal session cookies or sensitive data.
        Perform actions on behalf of users (e.g., CSRF).
        Deface the application or redirect users to malicious websites.

File: Secure Code Review/XSS/.Net/storexss.cs
using Microsoft.AspNetCore.Mvc;
using System.Collections.Generic;

namespace VulnerableApp.Controllers
{
    public class CommentsController : Controller
    {
        // In-memory storage for comments (simulating a database)
        private static List<string> comments = new List<string>();

        [HttpGet]
        public IActionResult Index()
        {
            // Display the comments
            ViewBag.Comments = comments;
            return View();
        }

        [HttpPost]
        public IActionResult Submit(string comment)
        {
            // Directly store the user input without validation or encoding
            comments.Add(comment);

            return RedirectToAction("Index");
        }
    }
}












Razor View (Index.cshtml):

<!DOCTYPE html>
<html>
<head>
    <title>Comments</title>
</head>
<body>
    <h1>Comments</h1>

    <form method="post" action="/Comments/Submit">
        <label for="comment">Add a Comment:</label>
        <input type="text" id="comment" name="comment" />
        <button type="submit">Submit</button>
    </form>

    <h2>Submitted Comments</h2>
    <ul>
        @foreach (var comment in ViewBag.Comments)
        {
            <li>@Html.Raw(comment)</li> <!-- Vulnerable to Stored XSS -->
        }
    </ul>
</body>
</html>



















...........................................................................................................................................
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////      
      


How This Code is Vulnerable

    Storing User Input Without Validation or Encoding:
        The application stores user input (comment) in the comments list without sanitization or encoding.
        Example: If an attacker submits <script>alert('XSS')</script>, it is stored as-is in the comments list.

    Unsafe Output Rendering:
        The Razor view uses @Html.Raw(comment) to display comments. This renders the HTML content directly in the browser.
        If the stored comment includes malicious JavaScript, it will be executed when the page is loaded.

    Example of Exploitation:
        An attacker submits <script>alert('XSS')</script> as a comment.
        When any user visits the comments page, the malicious script runs, showing an alert box or executing harmful code.

    Impact:
        Malicious scripts can:
            Steal user cookies or session tokens.
            Perform unauthorized actions on behalf of the user.
            Redirect users to malicious websites or deface the page.










